Fix transport boarding when master is on a transport (Zep/Boats)

Wishmaster117 opened the pull request:
Summary
This PR improves Follow related behaviour when the master is on a transport (zeppelin/boat). It makes follow actions safer and less disruptive by:

Detecting when the master is on a transport and handling boarding correctly
Avoiding teleport-under-floor issues by using a small positional offset when teleporting the bot near the master
Preventing movement conflicts between MoveSpline/MotionMaster and the transport driver by forcing a MotionMaster cleanup and MoveIdle after boarding
Clearing movement flags (forward / walking) after boarding so the bot does not remain in a walking/march state
Next-check delay after boarding to allow the server to update transport/position state

Before this change, bots get stuck when attempting to board

Fight the server-side transport movement because local MoveSpline/MotionMaster was still active
Repeatedly attempt movement on every follow tick while already a passenger, causing jitter and CPU/noise This PR reduces stuck/jitter cases, avoids conflicting movement commands, and makes boarding more robust.

**Key changes**
Check master->GetTransport() and handle three main cases:
If bot already passenger of same transport: stabilize (StopMoving, Clear(true), MoveIdle, StopMovingOnCurrentPos) and set a longer next-check delay; return false (no new movement in theory).
If bot passenger of another transport: do nothing (avoid conflicting behaviour).
If bot not a passenger of master transport: teleport bot near master (with offsets) and call Transport::AddPassenger(bot, true), 
then force:

bot->StopMoving()
bot->GetMotionMaster()->Clear(true)
bot->GetMotionMaster()->MoveIdle()

Remove movement flags MOVEMENTFLAG_FORWARD and MOVEMENTFLAG_WALKING
SetNextCheckDelay to random 1000–2500 ms
Log boarding with bot name, transport GUID and coordinates
Preserve earlier follow logic when master is not on a transport

Tests performed
Manual tests on a local server:

Master on boat/zeppelin -> bot teleports to a safe offset position and becomes a passenger without getting stuck
Bot already passenger on same transport -> bot no longer issues movement commands and stabilizes
Bot on a different transport -> no boarding attempt for master's transport (no interference)
Movement flags cleared after boarding; bot stops local movement and does not fight server transport movement

Now the bots follow their masters in the zeppelins and boats, although sometimes they move around a bit inside when the zeppelin starts (they must have smoked something bad).

Wishmaster117 committed:
Fix transport boarding when master is on a transport (Zep/Boats)

Wishmaster117 committed:
Opsiee

Wishmaster117 committed:
Update FollowActions.cpp

hermensbas commented on src/strategy/actions/FollowActions.cpp:
```diff
@@ -14,12 +14,57 @@
+            float y = master->GetPositionY() + offsetY;
+            float z = master->GetPositionZ() + offsetZ;
+
+            // Teleports to a position close to the master (world coordinates)
```
That being said, adding the map check is good enough for this commit for a working solution which awesome on itself. And maybe on top of that after this fix tested/merged we can think of more elegant way? One step at the time.

brighton-chi commented:
I mean, elevators are fucked up as is. The elevator boss in SSC was a meme back in the day, but bots... Holy hell, good thing they don't take fall damage.

Wishmaster117 commented:
> > > https://youtu.be/RZScS9rloMA
> > > Works as expected, map check for race conditions and possible distance check to master can make a bit more smoother with current teleport idea. Very nice!
> > > 
> > > * Pets dont teleport, but lets ignore pets pls for sake of complexity for this PR
> > > * Elevator seems fucked, not sure its consider to be transport object but prolly
> > 
> > 
> > To be honest, I'd completely forgotten about the elevators; I'll take a look.
> > However, for the pets, I tested it on the zeppelin in Orgrimmar and the ship in Stormwind, and they did get on them just fine, weird.
> 
> I think we could ignore pets for now for simplicity for the code, and when its merged, proved extend the functionality? I didnt think of elevators either until until i did :)

Elevators woks but need some adjust
https://youtu.be/4vhlC8EnLDY

Wishmaster117 committed:
Add Elevators handler

Wishmaster117 committed:
Clean up FollowActions.cpp by removing comments

Removed commented-out code related to transport handling.

Wishmaster117 committed:
Fix comment formatting in FollowActions.cpp

Wishmaster117 committed:
Fix comment formatting in FollowActions.cpp

Wishmaster117 commented:
Okay, I've commented the original code and added elevator handling. It works, but it still needs some adjustments for the elevators.
So we'll keep the code you deem most functional.

How to test different code:

Now the file have the V2 code.
To check with V1 code juste uncomment This part by removing /* at the start and */ at the end:
```
    /*
	// Transport (Zep + Boats) Fix
    // TODO: Smooth out bot boarding on transports (boats, zeppelins, tram).
	// TODO: Handle bot's pets
    Player* master = botAI->GetMaster();
    if (master && master->GetTransport())
    {
        Transport* transport = master->GetTransport();

        // Vérification de map : le bot doit être sur la même map que le maître
        if (bot->GetMap() != master->GetMap())
            return false;
	
        // If the bot is not already a passenger on this transport
        if (bot->GetTransport() != transport)
        {
            // Offset to avoid landing under the floor
            float offsetX = 1.0f, offsetY = 0.5f, offsetZ = 0.2f;
            float x = master->GetPositionX() + offsetX;
            float y = master->GetPositionY() + offsetY;
            float z = master->GetPositionZ() + offsetZ;

            // Teleports to a position close to the master (world coordinates)
            bot->TeleportTo(transport->GetMapId(), x, y, z, master->GetOrientation());

            // Add as a passenger (server-side)
            transport->AddPassenger(bot, true);

            // Force complete cleanup of the movement and flags to avoid
            // wrestling between MoveSpline/MotionMaster and the transport driver.
            bot->StopMoving();
            // MotionMaster current clear (true to force) then idle to stabilize
            bot->GetMotionMaster()->Clear(true);
            bot->GetMotionMaster()->MoveIdle();

            // Ensure that the march/mounted flags are no longer active
            bot->m_movementInfo.RemoveMovementFlag(MOVEMENTFLAG_FORWARD);
            bot->m_movementInfo.RemoveMovementFlag(MOVEMENTFLAG_WALKING);

            // Allow a slightly longer verification period; during this time
            // the server will update the transport status and relative positions.
            botAI->SetNextCheckDelay(urand(1000, 2500));

            LOG_DEBUG("playerbots", "Bot {} boarded transport {} near master {} at {:.2f},{:.2f},{:.2f}",
                     bot->GetName(), transport->GetGUID().GetCounter(), master->GetName(), x, y, z);

            return true;
        }
    } // End Transport (Zep + Boats) Fix
	*/
```

And add /* and */ at the start and the end of this part:

```
    // Unified Transport Handling (boats, zeppelins, elevators, platforms)
    Player* master = botAI->GetMaster();
    if (master && master->IsInWorld())
    {
        Map* map = master->GetMap();
        if (map && bot->GetMap() == map)
        {
            // Detect any kind of moving transport the master is currently standing on
            Transport* transport = master->GetTransport();

            // Some moving platforms/elevators do not expose a direct Transport* on the master.
            // In that case we fall back to a positional lookup on the map while the master
            // has the ONTRANSPORT movement flag.
            if (!transport && master->m_movementInfo.HasMovementFlag(MOVEMENTFLAG_ONTRANSPORT))
            {
                transport = map->GetTransportForPos(master->GetPhaseMask(),
                                                    master->GetPositionX(),
                                                    master->GetPositionY(),
                                                    master->GetPositionZ(),
                                                    master);
            }

            // If we found a transport and the bot is not yet attached, board it.
            if (transport && bot->GetTransport() != transport)
            {
                float const offsetX = 1.0f;
                float const offsetY = 0.5f;
                float const offsetZ = 0.2f;

                float const x = master->GetPositionX() + offsetX;
                float const y = master->GetPositionY() + offsetY;
                float const z = master->GetPositionZ() + offsetZ;

                // Teleport close to the master (world coordinates) and attach as passenger (server-side).
                bot->TeleportTo(transport->GetMapId(), x, y, z, master->GetOrientation());
                transport->AddPassenger(bot, true);

                // Reset movement state so the transport driver fully takes over.
                bot->StopMoving();
                bot->GetMotionMaster()->Clear(true);
                bot->GetMotionMaster()->MoveIdle();
                bot->m_movementInfo.RemoveMovementFlag(MOVEMENTFLAG_FORWARD);
                bot->m_movementInfo.RemoveMovementFlag(MOVEMENTFLAG_WALKING);

                botAI->SetNextCheckDelay(urand(1000, 2500));
                LOG_INFO("playerbots",
                          "Bot {} boarded transport {} near master {} at {:.2f},{:.2f},{:.2f}",
                          bot->GetName(), transport->GetGUID().GetCounter(), master->GetName(), x, y, z);
                return true;
            }
        }
    }
    // end unified transport handling
```

hermensbas commented:
I think the smart mount broke it, but that has been reverted. I will look into it after we managed the race conditions and crash log fixes you and blink made. Meanwhile i set on draft, the fix is great btw but give some time to play with it first.

Wishmaster117 committed:
Removal of TeleportTo

Removal of TeleportTo: as soon as bots are close enough (≤ 5 yards) to a master who is on a detected transport, they will be attached using AddPassenger and StopMovingOnCurrentPos().

Wishmaster117 commented:
> This is a hack that teleports them onto the transport. The idea should be to use regular systems for the bots. If player bots keeps making hacks for everything there will continue to be regressions

TeleportTo removed, still have difficulties to handle elevators....

Wishmaster117 committed:
Last adjustments

Boats => Green
Undercity Lift => Green
Zeplines => Green
Thunder Bluff Lifts => Green (so other lifts like this are ok)
Thousand Needles => Orange (Bots go to lift but don't stay in platform and reapear when  lift arrive to bottom)
Stormwind Tram => TODO

Wishmaster117 committed:
Fix codestyle

notOrrytrout commented:
This is on the radar for next test as I didn’t know how easy it is to check out a PR.

hermensbas committed:
fixes

hermensbas commented:
Testing and analysing, havent looked at the code yet. Just testing the behavior for now. Zepps and boats seems to work very nicely. Still testing elevators, platforms. Not ready to be merged yet.

Wishmaster117 committed:
Remove Elevators and Trams related code

Wishmaster117 committed:
Remove remaining platforms code in FollowActions

Celandriel committed:
Merge pull request #2128 from mod-playerbots/test-staging

Master update from Test-staging: Fix ObjectAccessor retrieval, optimize EquipActions, and implement RaidBossHelpers

hermensbas commented:
Will retest asap

hermensbas committed:
Revert "Convert PlayerBots tables to InnoDB (#2083)"

This reverts commit c86032f43b7bccb861c33769385db7a522e75513.

hermensbas committed:
Merge branch 'test-staging' into Fix-transport-boarding-when-master-is-on-a-transport-(Zep/Boats)

hermensbas committed:
Minor style

hermensbas committed:
not being used

hermensbas committed:
imports update

hermensbas commented:
rebased, cleaned up a bit , retested. Works good for boats and zepps.

Wishmaster117 commented on src/Bot/PlayerbotAI.cpp:
```diff
@@ -823,6 +823,11 @@ void PlayerbotAI::LeaveOrDisbandGroup()
     bot->GetSession()->QueuePacket(packet);
 }
 
+bool PlayerbotAI::IsMasterOnTransport()
```
Yes sorry, dead code i missed this one...
